### PR TITLE
*: support clustering

### PR DESCRIPTION
**Pull request by xhebox**
Signed-off-by: xhe <xw897002528@gmail.com>

### What problem does this PR solve?

Issue Number: ref #69

Problem Summary: Support a cluster of tiproxy nodes.

What is changed and how it works: Add five new cmd flags.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
```
n1: 3080,3081,6000
n2: 4080,4081,7000

 rm -rf work && ./bin/tiproxy --node_name=n1 --bootstrap_clusters "n2=127.0.0.1:4081" --config conf/proxy.yaml
rm -rf work2 && ./bin/tiproxy --node_name=n2 --bootstrap_clusters "n1=127.0.0.1:3081" --config conf2/proxy.yaml

After bootstraping

stop one node and restart it by ./bin/tiproxy  --config conf/proxy.yaml
```
- [ ] No code

Notable changes

- [ ] Has configuration change
- [ ] Has HTTP API interfaces change (Don't forget to [add the declarative for API](https://github.com/tikv/pd/blob/master/docs/development.md#updating-api-documentation))
- [ ] Has tiproxyctl change
- [ ] Other user behavior changes

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/server/server.go [hunk 1]
```diff
@@ -251,7 +252,28 @@ func (s *Server) Close() error {
...
+	}
+	if cfg.Cluster.BootstrapDurl != "" {
+		cnt++
+	}
+	if cfg.Cluster.BootstrapDdns != "" {
+		cnt++
+	}
+	if cnt > 1 {
+		return nil, errors.New("you can only pass one 'bootstrap_xxx' to bootstrap the node, or leave them empty to start a single-node cluster")
+	}
```
Comment (djshow832) [inline pkg/server/server.go:275]: Why make it so complicated, are other components also like this?

Comment (xhebox) [reply pkg/server/server.go:275]: This is common for etcd and other cloud native applications. Other pingcap components relies on the central PD(they only need PD).

PD only provide `bootstrapClusters` options, and a sugar option of `bootstrapCluster` by `-join xxx:2380` if you don't want to write all clusters by hands. I have no plan to include that sugar option in this PR.

Honestly, discovery or dns is more convenient.

To use `--initial-clusters`, you need to specify all nodes. To use `-join`, the cluster must be bootstraped already. Whatever you need to pass different options fot different PD and probably follow some orders.

Operator has implemented some very complex code.

Comment (djshow832) [reply pkg/server/server.go:275]: In my option, there's even no need to merge this PR, right? Proxies don't need to know each other, except for updating configurations. But updating configurations can be achieved by configuring the etcd address.

File: pkg/server/server.go [hunk 2]
```diff
@@ -251,7 +252,28 @@ func (s *Server) Close() error {
 	return errors.Collect(ErrCloseServer, errs...)
 }
 
-func buildEtcd(cfg *config.Config, logger *zap.Logger, pubAddr string, engine *gin.Engine) (srv *embed.Etcd, err error) {
+func buildEtcd(cfg *config.Config, logger *zap.Logger, engine *gin.Engine) (srv *embed.Etcd, err error) {
+	if cfg.Cluster.ClusterName == "" {
+		return nil, errors.New("cluster_name can not be empty")
+	}
+	if cfg.Cluster.NodeName == "" {
+		cfg.Cluster.NodeName = fmt.Sprintf("%s-%d", cfg.Cluster.ClusterName, time.Now().UnixMicro())
```
Comment (djshow832) [inline pkg/server/server.go:260]: Does PD also generate it like this?

Comment (xhebox) [reply pkg/server/server.go:260]: No, but in a smilar fashion. It needs to be unique in the cluster.

File: pkg/server/server.go (hunk 1 above)
Comment (xhebox) [reply pkg/server/server.go:275]: Only if you want to remove etcd from tiproxy, and relies on an external ETCD. I think it makes things complicated. If you have PD, you don't want another etcd. If you use tiproxy with static servers, you need another etcd.

File: pkg/server/server.go (hunk 2 above)
Comment (djshow832) [reply pkg/server/server.go:260]: Can we also generate it like that? A hostname is better than a random timestamp.

**Review by djshow832 — APPROVED** [review]

**Pull request merged**